filters/include/pcl/filters/lambda_filter_indices.h

```diff
@@ -0,0 +1,166 @@
+ *
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder(s) nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ * $Id$
+ *
+ */
+
+#pragma once
+
+#include <pcl/filters/filter_indices.h>
+#include <pcl/pcl_macros.h>
+#include <pcl/type_traits.h> // for is_invocable
+
+#ifndef __cpp_lib_is_invocable
+#include <boost/hof/is_invocable.hpp>
+#endif
+
+#include <type_traits>
+
+#include <utility>
+
+namespace pcl {
+namespace detail {
+template <typename PointT, typename Function>
+using PointFilterLambda =
+    std::enable_if_t<pcl::is_invocable_r_v<bool,
+                                           Function,
+                                           const pcl::remove_cvref_t<PointT>&,
+                                           pcl::index_t>,
+                     bool>;
+
+// can't use this for SFINAE since Derived isn't properly defined
+// but this can be used after the class definition to test it
+/*
+template <class Base, class Derived>
+constexpr auto IsValidLambdaFilter = std::enable_if_t<
+    std::is_base_of<Base, Derived>::value &&
+        pcl::is_invocable_v<std::declval<Derived>().get_lambda, void> &&
+        std::is_same<PointFilterLambda<std::declval<Derived>().get_lambda()>,
+bool>, bool>;
+*/
+
+/**
+ * \brief LambdaFilterIndices filters point clouds and indices based on a
+ * function pointer passed to filter command \ingroup filters
+ */
+template <typename PointT, typename Derived>
+struct LambdaFilterIndicesImpl : public FilterIndices<PointT> {
```

Comment:
What's the motivation for creating `LambdaFilterIndicesImpl` and `LambdaFilterIndices` as separate classes?

Reply:
One is a storage class and another is logic class. This potentially allows future support of immovable functors via specializations of the storage class, without disrupting/duplicating the core logic.

Reply:
An example would be detecting if the functor is a lambda and convertible to a `bool(*)(PointT, index_t)` and then adding [[no_unique_address]] to the variable (because the lambda class would empty)